ground/src/main/java/com/google/android/ground/ui/datacollection/TaskSequenceHandlerImpl.kt

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.android.ground.ui.datacollection
+
+import com.google.android.ground.model.submission.TaskData
+import com.google.android.ground.model.task.Task
+
+class TaskSequenceHandlerImpl(
+  private val tasks: List<Task>,
+  private val shouldIncludeTask:
+    (task: Task, taskValueOverride: Pair<String, TaskData?>?) -> Boolean,
+) : TaskSequenceHandler {
+
+  init {
+    if (tasks.isEmpty()) {
+      error("Can't generate sequence for empty task list")
+    }
+  }
+
+  private fun checkInvalidTaskId(taskId: String) {
+    if (taskId.isBlank()) {
+      error("Task ID can't be blank")
+    }
+  }
+
+  private fun checkInvalidIndex(taskId: String, index: Int) {
+    checkInvalidTaskId(taskId)
+    if (index < 0) {
+      error("Task $taskId not found in task list")
+    }
+  }
+
+  override fun getTaskSequence(taskValueOverride: Pair<String, TaskData?>?): Sequence<Task> =
+    tasks.filter { task -> shouldIncludeTask(task, taskValueOverride) }.asSequence()
+
+  override fun isFirstPosition(taskId: String): Boolean {
+    checkInvalidTaskId(taskId)
+    return taskId == getTaskSequence().first().id
+  }
+
+  override fun isLastPosition(taskId: String): Boolean {
+    checkInvalidTaskId(taskId)
+    return taskId == getTaskSequence().last().id
+  }
+
+  override fun isLastPosition(taskId: String, value: TaskData?): Boolean {
+    checkInvalidTaskId(taskId)
+    return taskId == getTaskSequence(taskValueOverride = taskId to value).last().id
+  }
+
+  override fun getPreviousTask(taskId: String): String {
+    checkInvalidTaskId(taskId)
+    return getTaskSequence().elementAt(getRelativePosition(taskId) - 1).id
+  }
+
+  override fun getNextTask(taskId: String): String {
+    checkInvalidTaskId(taskId)
+    return getTaskSequence().elementAt(getRelativePosition(taskId) + 1).id
+  }
+
+  override fun getAbsolutePosition(taskId: String): Int {
+    val index = tasks.indexOfFirst { it.id == taskId }
+    checkInvalidIndex(taskId, index)
+    return index
+  }
+
+  override fun getRelativePosition(taskId: String): Int {
+    val index = getTaskSequence().indexOfFirst { it.id == taskId }
+    checkInvalidIndex(taskId, index)
```

Comment:
Consider returning a sentinel `-1` value instead of throwing. This is in keeping with the semantics of `indexOfFirst` and would allow you to make this just `return getTaskSequence().indexOfFirst { it.id == taskId }`.

I would generally avoid throwing unless we need to really flag checking for a task no in the collection as an exceptional case (which might be the correct thing here, I trust your call on this)

Reply:
Up until this point, we were assuming -1 means first task. IMO, that added fragility to the code as instead of failing the application, we were sometimes jumping tasks if something couldn't be found. So, to identify all those default cases, I added the task id empty check and index out of bounds check. 

Let me re-evaluate if we can live without throwing the errors. Thanks again!